src/commands/update.ts

```diff
@@ -0,0 +1,41 @@
+import { SlashCommandBuilder } from 'discord.js';
+import { exec } from 'node:child_process';
+
+const info = new SlashCommandBuilder()
+	.setName('update')
+	.setDescription('Pulls the latest changes and restarts the bot')
+	.setDefaultMemberPermissions('0');
+
+export const update: GlobalCommand = {
+	info,
+	requiresGuild: false,
+	async execute({ replyPrivately, user, interaction }) {
+		const admin_ids = process.env['ADMINISTRATORS']?.split(',');
+		if (!admin_ids) {
+			// TODO: make this a UserMessageException
+			throw new Error('There is no ADMINISTRATORS variable. You must set ADMINISTRATORS in .env');
+		}
+
+		if (!admin_ids.includes(user.id)) {
+			// TODO: make this a UserMessageException
+			throw new Error(
+				'You do not have permission to perform this command. Contact the bot administrator.'
+			);
+		}
+
+		await replyPrivately('Updating...');
+		exec('npm run update', err => {
```

Comment:
What consequences might there be to this command being run several times in quick succession, by one or more users?

Reply:
There are several ways to handle this. In general, each invocation needs to check that it is singleton, and if it is not, abort execution. Some ideas come to mind:
- A global `boolean` variable that declares whether this command is being invoked. To avoid race conditions, that variable should be set early and checked frequently, so as to back out early in case of spoopy behavior.
- A global `number` or `bigint` variable that declares the number of running instances. Similar to the global `boolean` idea, with added information for debugging. ("Why was this thing invoked _3_ times in a row?")

We might consider storing this info at the top level of the module, letting Node disambiguate invocations within its process. A more robust model might store the info in the database, so as to be resilient to multiple Node process firing at once (slightly more likely since we now defer the runtime execution to PM2).

This is only a simple "do an update" command, but its self-surgery nature demands careful thought to risk management.

Reply:
This is a good point. The solution I've added counts the number of invocations, and based on how JS's event loop works, we shouldn't run into any race conditions. As for cases where there are mutliple instances of the bot running, that shouldn't happen since pm2 would try to give them the same process name. If it were to happen though, nothing short of like an inter-process semaphore would work (is that a thing?), so I don't think it's worth hooking it up to the db.

As far as what the consequences are: the bot could fail to restart, possibly the `dist` code gets corrupted in some way, in which case manual intervention would be required. Considering though that this command is mainly a convenience so I (and future bot admins) don't need to manually update it every time, even if it occassionally fails that's a win. With this new check, I don't know of a way it could fail (short of the bot admin starting another bot instance outside pm2 I guess), but still